Do not use exprt::move_to_operands as it is marked as deprecated

tautschnig opened the pull request:
In particular, the use in std_code.h triggers a series of warnings every time
this file is included.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- n/a My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 141b5e598cb7cde8e9a1ac7c21eb3ab23f22eacc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/89882763

smowton commented on src/util/std_code.h:
```diff
@@ -139,7 +139,9 @@ class code_blockt:public codet
   void move(codet &code)
   {
-    move_to_operands(code);
+    code_operandst &stmts = statements();
```
This seems obviously worse. Given move semantics are fairly widely used in the codebase and not deprecated, and `irept::swap` is used to fake move semantics, why not just remove the tag from `move_to_operands`?

tautschnig replied:
What about making `move` take an rvalue reference? (But also I think getting rid of `exprt::move_to_operands` in here is actually a good idea, because eventually a `codet` should no longer be an `exprt`.)

smowton replied:
In the linked PR I've gone for an `add_to_operands` that takes either kind of reference and behaves appropriately. Re: codet / exprt, I'd say change that when the hierarchy is changed

smowton commented:
In https://github.com/diffblue/cbmc/pull/3067 the solution arrived at was to have a rvalue-ref version of `add`. Rather oddly this was implemented by creating rvalue-ref overloads of `copy_to_operands` (https://github.com/diffblue/cbmc/blob/develop/src/util/expr.h#L126), so we end up writing `some_expr.copy_to_operands(std::move(other_expr))`.

I'd suggest simply renaming `copy_to_operands(exprt &&)` as `move_to_operands(exprt &&)` or `add_to_operands(exprt &&)` and use that function to implement this PR.

smowton commented:
PR here: https://github.com/diffblue/cbmc/pull/3254

tautschnig committed:
Do not use exprt::move_to_operands as it is marked as deprecated

expr.cpp should not be using methods marked as deprecated in its own header
file. This change furthermore increases type safety as the constructor of
`not_exprt` will check that the type of the operand is ID_bool.

tautschnig commented:
I have removed the changes to `codet` from this commit/PR, I will do them separately based on the work that @smowton has done in the meantime. Hopefully this is now no longer controversial.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: ec5e0bea58b1ac851de07e066fc137afb5750d4f).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/89942582
